group.c

```diff
@@ -77,26 +78,31 @@ static int empty_group(struct Group *g)
   return !g->as && !g->rs;
 }
 
-void mutt_group_context_add(struct GroupContext **ctx, struct Group *group)
+void mutt_group_context_add(struct GroupContextHead *head, struct Group *group)
 {
-  for (; *ctx; ctx = &((*ctx)->next))
+  struct GroupContext *np = NULL;
+  STAILQ_FOREACH(np, head, entries)
   {
-    if ((*ctx)->g == group)
+    if (np->g == group)
       return;
   }
-
-  *ctx = mutt_mem_calloc(1, sizeof(struct GroupContext));
-  (*ctx)->g = group;
+  np = mutt_mem_calloc(1, sizeof(struct GroupContext));
+  np->g = group;
+  STAILQ_INIT(head);
```

Comment:
you want to insert the newly created node here, not reinitialize the whole list.